src/Common/src/Internal/Runtime/EEType.cs

```diff
@@ -1415,6 +1467,84 @@ internal unsafe struct EETypeRef
         }
     }
 
+    // Wrapper around pointers
+    [StructLayout(LayoutKind.Sequential)]
+    internal unsafe struct Pointer<T> where T : unmanaged
+    {
+        private readonly T* _value;
+
+        public T* Value
+        {
+            get
+            {
+                return _value;
+            }
+        }
+    }
+
+    // Wrapper around pointers that might be indirected through IAT
+    [StructLayout(LayoutKind.Sequential)]
+    internal unsafe struct IatAwarePointer<T> where T : unmanaged
+    {
+        private readonly T* _value;
+
+        public T* Value
+        {
+            get
+            {
+                if (((int)_value & IndirectionConstants.IndirectionCellPointer) == 0)
+                    return _value;
+                return *(T**)((byte*)_value - IndirectionConstants.IndirectionCellPointer);
+            }
+        }
+    }
+
+    // Wrapper around relative pointers
+    [StructLayout(LayoutKind.Sequential)]
+    internal unsafe struct RelativePointer<T> where T : unmanaged
+    {
+        private readonly int _value;
+
+        public T* Value
+        {
+            get
+            {
+                fixed (int* pValue = &_value)
+                {
+                    return (T*)((byte*)pValue + _value);
+                }
+            }
+        }
+    }
+
+    // Wrapper around relative pointers that might be indirected through IAT
+    [StructLayout(LayoutKind.Sequential)]
+    internal unsafe struct IatAwareRelativePointer<T> where T : unmanaged
+    {
+        private readonly int _value;
+
+        public T* Value
+        {
+            get
+            {
+                if ((_value & IndirectionConstants.IndirectionCellPointer) == 0)
```

Comment:
On Project N, we typically use RVA relocs for this with the highest bit set. In order to preserve sanity, I decided to go with relative pointers on all ABIs. The drawback is we can no longer use highest bit because that one means a negative offset. I'm using lowest bit instead, under the assumption that we'll never use this to generate relocs between things that are not at least 2-byte aligned.

Reply:
We need to be careful with the lowest bit for ARM32 thumb code reasons, but as long as there are never code pointers, I think we're largely ok here. You'll need to audit that the alignment of the data structures you generate is higher than 1 though. I think you should be able to put some sort of assertion to that effect into the system. We do have the ability to emit data structures with 1 byte alignment, but I don't know if we use it much.

---
In reply to: [199278989](https://github.com/dotnet/corert/pull/6035#discussion_r199278989) [](ancestors = 199278989)